File: packages/ml/src/features.ts
import type { CausalEdge } from '@blamr/types';
import { cosineSimilarity } from '@blamr/semantic';

/** Fixed hop feature dimension — must match training pipeline. */
export const HOP_FEATURE_DIM = 24;

const INFLATION_THRESHOLD = 0.15;

export interface HopFeatureContext {
  edges: CausalEdge[];
  hopIndex: number;
  cosInOut?: number;
  cosGoalOut?: number;
}

function callTypeOneHot(callType: string): [number, number, number, number] {
  const t = callType.toLowerCase();
  if (t.includes('tool')) return [0, 1, 0, 0];
  if (t.includes('mcp')) return [0, 0, 1, 0];
  if (t.includes('vision')) return [0, 0, 0, 1];
  return [1, 0, 0, 0];
}

function logNorm(value: number, scale: number): number {
  return Math.log1p(Math.max(0, value)) / scale;
}

/** Build a fixed-size feature vector for one causal hop. */
export function extractHopFeatures(ctx: HopFeatureContext): Float64Array {
  const sorted = [...ctx.edges].sort((a, b) => a.hop_index - b.hop_index);
  const edge = sorted.find((e) => e.hop_index === ctx.hopIndex);
  if (!edge) return new Float64Array(HOP_FEATURE_DIM);

  const n = sorted.length;
  const idx = sorted.findIndex((e) => e.hop_index === ctx.hopIndex);
  const prev = idx > 0 ? sorted[idx - 1] : null;

  const confDrop = Math.max(0, edge.confidence_in - edge.confidence_out);
  const inflation = Math.max(0, edge.confidence_out - edge.confidence_in - INFLATION_THRESHOLD);
  const intentHarm = Math.max(0, -edge.intent_delta);
  const tokens = edge.tokens_in + edge.tokens_out;

  const prevIntentHarm = prev ? Math.max(0, -prev.intent_delta) : 0;
  const prevConfDrop = prev ? Math.max(0, prev.confidence_in - prev.confidence_out) : 0;

  const cosInOut = ctx.cosInOut ?? 0.5;
  const cosGoalOut = ctx.cosGoalOut ?? 0.5;

  const vec = new Float64Array(HOP_FEATURE_DIM);
  vec[0] = edge.confidence_in;
  vec[1] = edge.confidence_out;
  vec[2] = edge.intent_delta;
  vec[3] = edge.influence_score;
  vec[4] = confDrop;
  vec[5] = inflation;
  vec[6] = intentHarm;
  vec[7] = n > 1 ? idx / (n - 1) : 0;
  vec[8] = logNorm(tokens, 10);
  vec[9] = logNorm(edge.latency_ms, 10);
  vec[10] = logNorm(edge.cost_usd * 1000, 5);
  const [llm, tool, mcp, vision] = callTypeOneHot(edge.call_type);
  vec[11] = llm;
  vec[12] = tool;
  vec[13] = mcp;
  vec[14] = vision;
  vec[15] = cosInOut;
  vec[16] = cosGoalOut;
  vec[17] = edge.input_preview?.trim() ? 1 : 0;
  vec[18] = edge.output_preview?.trim() ? 1 : 0;
  vec[19] = idx === 0 ? 1 : 0;
  vec[20] = idx === n - 1 ? 1 : 0;
  vec[21] = prevIntentHarm;
  vec[22] = prevConfDrop;
  vec[23] = edge.confidence_out * edge.influence_score;

  return vec;
}

/** Cosine similarities for embedding-enriched features. */
export function embeddingSimilarities(
  inputVec: number[] | undefined,
  outputVec: number[] | undefined,
  goalVec: number[] | undefined,
): { cosInOut: number; cosGoalOut: number } {
  let cosInOut = 0.5;
  let cosGoalOut = 0.5;
  if (inputVec && outputVec) cosInOut = cosineSimilarity(inputVec, outputVec);
  if (goalVec && outputVec) cosGoalOut = cosineSimilarity(goalVec, outputVec);
  return { cosInOut, cosGoalOut };
}

export const AGENT_FEATURE_DIM = 8;

/** Per-agent features aggregated from hop-level ML outputs for ranker. */
export function extractAgentFeatures(
  agent: string,
  edges: CausalEdge[],
  hopDriftScores: Map<number, number>,
  hopIntentHarm: Map<number, number>,
): Float64Array {
  const fromHops = edges.filter((e) => e.from_agent === agent);
  const vec = new Float64Array(AGENT_FEATURE_DIM);

  if (fromHops.length === 0) return vec;

  let maxDrift = 0;
  let sumDriftInfluence = 0;
  let maxIntentHarm = 0;
  let maxConfDrop = 0;
  let inflationFlag = 0;
  let sumInfluence = 0;

  for (const e of fromHops) {
    const drift = hopDriftScores.get(e.hop_index) ?? 0;
    const harm = hopIntentHarm.get(e.hop_index) ?? Math.max(0, -e.intent_delta);
    maxDrift = Math.max(maxDrift, drift);
    sumDriftInfluence += drift * e.influence_score;
    maxIntentHarm = Math.max(maxIntentHarm, harm);
    maxConfDrop = Math.max(maxConfDrop, Math.max(0, e.confidence_in - e.confidence_out));
    if (e.confidence_out - e.confidence_in > INFLATION_THRESHOLD) inflationFlag = 1;
    sumInfluence += e.influence_score;
  }

  vec[0] = maxDrift;
  vec[1] = sumDriftInfluence;
  vec[2] = maxIntentHarm;
  vec[3] = maxConfDrop;
  vec[4] = inflationFlag;
  vec[5] = fromHops.length / Math.max(edges.length, 1);
  vec[6] = sumInfluence / fromHops.length;
  vec[7] = fromHops[0]?.hop_index === Math.min(...edges.map((e) => e.hop_index)) ? 1 : 0;

  return vec;
}

File: training/features.py
"""Feature extraction mirroring packages/ml/src/features.ts (HOP_FEATURE_DIM=24)."""
from __future__ import annotations

import math
from typing import Any

HOP_FEATURE_DIM = 24
AGENT_FEATURE_DIM = 8
INFLATION_THRESHOLD = 0.15

DRIFT_CLASSES = [
    "none",
    "domain_mismatch",
    "retrieval_miss",
    "severity_underrate",
    "confidence_inflation",
    "propagation",
    "format_error",
]


def call_type_one_hot(call_type: str) -> list[float]:
    t = call_type.lower()
    if "tool" in t:
        return [0.0, 1.0, 0.0, 0.0]
    if "mcp" in t:
        return [0.0, 0.0, 1.0, 0.0]
    if "vision" in t:
        return [0.0, 0.0, 0.0, 1.0]
    return [1.0, 0.0, 0.0, 0.0]


def log_norm(value: float, scale: float) -> float:
    return math.log1p(max(0.0, value)) / scale


def extract_hop_features(edge: dict[str, Any], idx: int, n: int, prev: dict[str, Any] | None) -> list[float]:
    ci = float(edge["confidence_in"])
    co = float(edge["confidence_out"])
    intent_delta = float(edge["intent_delta"])
    influence = float(edge["influence_score"])
    conf_drop = max(0.0, ci - co)
    inflation = max(0.0, co - ci - INFLATION_THRESHOLD)
    intent_harm = max(0.0, -intent_delta)
    tokens = int(edge.get("tokens_in", 0)) + int(edge.get("tokens_out", 0))
    prev_intent = max(0.0, -float(prev["intent_delta"])) if prev else 0.0
    prev_drop = max(0.0, float(prev["confidence_in"]) - float(prev["confidence_out"])) if prev else 0.0
    llm, tool, mcp, vision = call_type_one_hot(str(edge.get("call_type", "LLM call")))
    cos_in_out = float(edge.get("cos_in_out", 0.5))
    cos_goal_out = float(edge.get("cos_goal_out", 0.5))

    return [
        ci,
        co,
        intent_delta,
        influence,
        conf_drop,
        inflation,
        intent_harm,
        idx / (n - 1) if n > 1 else 0.0,
        log_norm(tokens, 10),
        log_norm(float(edge.get("latency_ms", 0)), 10),
        log_norm(float(edge.get("cost_usd", 0)) * 1000, 5),
        llm,
        tool,
        mcp,
        vision,
        cos_in_out,
        cos_goal_out,
        1.0 if edge.get("input_preview") else 0.0,
        1.0 if edge.get("output_preview") else 0.0,
        1.0 if idx == 0 else 0.0,
        1.0 if idx == n - 1 else 0.0,
        prev_intent,
        prev_drop,
        co * influence,
    ]


def extract_agent_features(
    agent: str,
    edges: list[dict[str, Any]],
    hop_drift: dict[int, float],
) -> list[float]:
    from_hops = [e for e in edges if e["from_agent"] == agent]
    if not from_hops:
        return [0.0] * AGENT_FEATURE_DIM

    max_drift = max(hop_drift.get(e["hop_index"], 0.0) for e in from_hops)
    sum_drift_inf = sum(hop_drift.get(e["hop_index"], 0.0) * e["influence_score"] for e in from_hops)
    max_harm = max(max(0.0, -e["intent_delta"]) for e in from_hops)
    max_drop = max(max(0.0, e["confidence_in"] - e["confidence_out"]) for e in from_hops)
    inflation = 1.0 if any(e["confidence_out"] - e["confidence_in"] > INFLATION_THRESHOLD for e in from_hops) else 0.0
    sum_inf = sum(e["influence_score"] for e in from_hops)
    min_hop = min(e["hop_index"] for e in edges)

    return [
        max_drift,
        sum_drift_inf,
        max_harm,
        max_drop,
        inflation,
        len(from_hops) / max(len(edges), 1),
        sum_inf / len(from_hops),
        1.0 if from_hops[0]["hop_index"] == min_hop else 0.0,
    ]
